### PR TITLE
[WIP] Logs support for sumologic k8s collection

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:** <Describe what has changed. 
Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.>

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** < Describe what testing was performed and which tests were added.>

**Documentation:** < Describe the documentation added.>

### CODE REVIEW

**Comment by pmm-sumo** [comment]
That's great work. Few comments:
* I would consider splitting the PR for adding logs to k8processor and opening it against the main OpenTelemetry Collector Contrib repo - it should not impact our custom extensions much and it's already on the roadmap anyway
* I am not sure if that would change anything, but it's plausible we'll add other pillars to the Sumo Exporter (metrics, traces) - even those will be sent simply via OTLP and the current `source` processor code would be moved there as well

File: processor/fluentbitk8sprocessor/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## <a name="fluentbitk8sprocessor"></a>Fluent bit k8s rocessor
```
Comment (pmm-sumo) [inline processor/fluentbitk8sprocessor/README.md:1]: Is this processor specific to sumologic-kubernetes-collection?

Comment (sumo-drosiek) [reply processor/fluentbitk8sprocessor/README.md:1]: So in order to perform metadata on the logs from the fluentbit from the k8s we need to apply that plugin. I believe it can be useful for the community too, eventually the functionality can be moved to the fluentforwarder, but I think it's better as separated processor

**Comment by sumo-drosiek** [comment]
> 
>     * I would consider splitting the PR for adding logs to k8processor and opening it against the main OpenTelemetry Collector Contrib repo - it should not impact our custom extensions much and it's already on the roadmap anyway

I will take care of that

>     * I am not sure if that would change anything, but it's plausible we'll add other pillars to the Sumo Exporter (metrics, traces) - even those will be sent simply via OTLP and the current `source` processor code would be moved there as well

I'm going to add metrics to the sumologic exporter in the metrics pipeline PR, and regarding traces I don't know 🤷

File: processor/fluentbitk8sprocessor/README.md (hunk 1 above)
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/README.md:1]: 1) 

```suggestion
## <a name="fluentbitk8sprocessor"></a>Fluent bit k8s processor
```

2) 
> `<a name="fluentbitk8sprocessor"></a>`

Do we need this?

File: processor/fluentbitk8sprocessor/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+Processor which convert fluentd tag to the k8s metadata:
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/README.md:3]:
```suggestion
Processor which converts fluentd tags to the following k8s metadata:
```

File: processor/fluentbitk8sprocessor/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+ - `pod_name`
+ - `namespace`
+ - `container_name`
+ - `docker_id`
+
+ This plugin is required to correct work of k8s metadata
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/README.md:9]:
```suggestion
 This plugin is required for correct work of k8s metadata
```

File: exporter/sumologicexporter/go.mod [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/exporter/sumologicexporter
+
+go 1.14
+
+replace go.opentelemetry.io/collector => github.com/SumoLogic/opentelemetry-collector v0.8.0
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/go.mod:5]: Hmm why did you need to add this? I'm almost certain we won't be able to merge this like this to upstream

File: processor/fluentbitk8sprocessor/Makefile [hunk 5]
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/Makefile:1]: End line character at the end

File: processor/fluentbitk8sprocessor/processor.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+func (fbkp *fluentbitk8sprocessor) ExtractMetadata(attributes pdata.AttributeMap) {
+	attr, exists := attributes.Get("fluent.tag")
+
+	if exists && attr.Type() == pdata.AttributeValueSTRING {
+		tag := attr.StringVal()
+		regex := regexp.MustCompile(`\.containers\.([^_]+)_([^_]+)_(.+)-([a-z0-9]{64})\.log$`)
+		details := regex.FindAllStringSubmatch(tag, -1)
+
+		if len(details) == 1 && len(details[0]) == 5 {
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/processor.go:63]: Maybe comment here why we're expecting `len(details[0]) == 5`?

File: processor/fluentbitk8sprocessor/processor.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+			attributes.InsertString("k8s.pod.name", details[0][1])
+			attributes.InsertString("pod_name", details[0][1])
+			attributes.InsertString("namespace", details[0][2])
+			attributes.InsertString("container_name", details[0][3])
+			attributes.InsertString("docker_id", details[0][4])
+		}
+	}
+}
+
+// ConsumeLogs process logs and convberts fluentbit k8s tag to metadata
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/processor.go:73]:
```suggestion
// ConsumeLogs processes logs and converts fluentbit k8s tags to metadata
```

File: processor/fluentbitk8sprocessor/processor.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+func (fbkp *fluentbitk8sprocessor) ConsumeLogs(ctx context.Context, logs pdata.Logs) error {
+	for i := 0; i < logs.ResourceLogs().Len(); i++ {
+		resource := logs.ResourceLogs().At(i)
+
+		// ToDo: fluent bit tag can potantially by in resource attributes
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/processor.go:78]:
```suggestion
		// ToDo: fluent bit tag can potantially be in resource attributes
```

File: exporter/sumologicexporter/exporter.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	attributes.Sort().ForEach(func(k string, v pdata.AttributeValue) {
+		buf.WriteString(k)
+		buf.WriteString("=")
+		buf.WriteString(v.StringVal())
+		buf.WriteString(", ")
+	})
+	return buf.String()
+}
+
+func (se *sumologicexporter) Send(buffer []pdata.LogRecord, fields string) {
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/exporter.go:110]: Add doc comments for all exported methods like this one

File: exporter/sumologicexporter/exporter.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	client := &http.Client{}
+	body := bytes.Buffer{}
```
Comment (pmalek-sumo) [inline exporter/sumologicexporter/exporter.go:112]: Using [strings.Builder](https://golang.org/pkg/strings/#Builder) might be better for this - we're operating on strings anyway, not on bytes

File: processor/fluentbitk8sprocessor/factory.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+
+package fluentbitk8sprocessor
+
+import (
+	"go.opentelemetry.io/collector/processor/processorhelper"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/component"
+	"context"
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/factory.go:22]: I'd stick to the following format (as it's done in the other files) + sort the lines alphabetically

```suggestion
	"context"

	"go.opentelemetry.io/collector/component"
	"go.opentelemetry.io/collector/config/configmodels"**
	"go.opentelemetry.io/collector/consumer"
	"go.opentelemetry.io/collector/processor/processorhelper"
```

File: processor/fluentbitk8sprocessor/go.mod [hunk 12]
```diff
@@ -0,0 +1,11 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/processor/fluentbitk8sprocessor
+
+go 1.14
+
+replace go.opentelemetry.io/collector => github.com/SumoLogic/opentelemetry-collector v0.8.0
```
Comment (pmalek-sumo) [inline processor/fluentbitk8sprocessor/go.mod:5]: Same possible import problem here

File: exporter/sumologicexporter/go.mod (hunk 4 above)
Comment (sumo-drosiek) [reply exporter/sumologicexporter/go.mod:5]: because we are using our modified version of the opentelemetry-collector (in general). I think this exporter doesn't need this. Also I'm going to create more reliable PRs for the upstream

File: processor/sourceprocessor/source_processor.go [hunk 13]
```diff
@@ -424,3 +425,69 @@ func (f *attributeFiller) resourceLabelValues(atts *pdata.AttributeMap) []interf
...
+
+				if !res.IsNil() {
+					atts := res.Attributes()
+
+					stp.enrichPodName(&atts)
+					stp.fillOtherMeta(atts)
+
+					filledAnySource = stp.sourceHostFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceHostSpecialAnnotation), stp.keys) || filledAnySource
+					filledAnySource = stp.sourceCategoryFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceCategorySpecialAnnotation), stp.keys) || filledAnySource
+					filledAnySource = stp.sourceNameFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceNameSpecialAnnotation), stp.keys) || filledAnySource
```
Comment (pmalek-sumo) [inline processor/sourceprocessor/source_processor.go:480]: `filledAnySource` seems to be unused?

File: processor/fluentbitk8sprocessor/README.md (hunk 1 above)
Comment (sumo-drosiek) [reply processor/fluentbitk8sprocessor/README.md:1]: > Do we need this?

I use this convention as it's used across the otelcol repository

**Comment by sumo-drosiek** [comment]
Closing due to re-write
